gateway/src/node/client_handling/websocket/connection_handler/ecash/mod.rs

```diff
@@ -0,0 +1,481 @@
+        Ok(())
+    }
+
+    pub async fn check_double_spend(&self, serial_number_bs58: &String) -> bool {
+        if let Some(double_spend_detector) = &self.double_spend_detector {
+            double_spend_detector.check(serial_number_bs58).await
+        } else {
+            false
+        }
+    }
+
+    pub async fn post_credential(
+        &self,
+        api_clients: &[CoconutApiClient],
+        credential: CredentialSpendingRequest,
+    ) -> Result<(), RequestHandlingError> {
+        self.cred_sender
+            .unbounded_send(PendingCredential::new(
+                credential.clone(),
+                self.nyxd_client.read().await.address(),
+                api_clients
+                    .iter()
+                    .map(|client| client.api_client.clone())
+                    .collect(),
+            ))
+            .map_err(|_| RequestHandlingError::InternalError)
+    }
+
+    pub async fn spend_online_credential(
+        &self,
+        api_clients: &[CoconutApiClient],
+        credential: &CredentialSpendingRequest,
+    ) -> Result<(), RequestHandlingError> {
+        let serial_number = credential.data.payment.serial_number_bs58();
+        let proposal_id = match credential.data.typ {
+            CredentialType::Voucher => unimplemented!(),
+            CredentialType::TicketBook => {
+                let res = self
+                    .nyxd_client
+                    .write()
+                    .await
+                    .prepare_credential(serial_number, self.address.to_string(), None)
+                    .await?;
+                let proposal_id = find_attribute(&res.logs, "wasm", BANDWIDTH_PROPOSAL_ID)
+                    .ok_or(RequestHandlingError::ProposalIdError {
+                        reason: String::from("proposal id not found"),
+                    })?
+                    .value
+                    .parse::<u64>()
+                    .map_err(|_| RequestHandlingError::ProposalIdError {
+                        reason: String::from("proposal id could not be parsed to u64"),
+                    })?;
+
+                let proposal = self
+                    .nyxd_client
+                    .read()
+                    .await
+                    .query_proposal(proposal_id)
+                    .await?;
+                if !credential.matches_serial_number(&proposal.description)? {
+                    return Err(RequestHandlingError::ProposalIdError {
+                        reason: String::from("proposal has different serial number"),
+                    });
+                }
+                Some(proposal_id)
+            }
+            CredentialType::FreePass => None,
+        };
+
+        let req = nym_api_requests::coconut::VerifyEcashCredentialBody::new(
+            credential.data.clone(),
+            self.address.clone(),
+            proposal_id,
+        );
+        for client in api_clients {
+            let ret = client.api_client.verify_online_credential(&req).await;
```

Comment:
do we have to verify it with **ALL** apis?
also, we might want to do it in parallel. look at `for_each_concurrent`

Reply:
`threshold` should suffice, though all of them would be better